pkg/scheduler/queue/tree_queue.go

```diff
@@ -4,45 +4,138 @@ package queue
+func (t *TreeQueue) Dequeue() (QueuePath, any) {
+	path, v := t.rootNode.dequeue()
+	// The returned node dequeue path includes the root node; exclude
+	// this so that the return path can be used if needed to enqueue.
+	return path[1:], v
+}
+
+// EnqueueBackByPath enqueues an item in the back of the local queue of the node
+// located at a given path through the tree; nodes for the path are created as needed.
+//
+// path is relative to the root node; providing a QueuePath beginning with "root"
+// will create a child node of the root node which is also named "root."
+func (t *TreeQueue) EnqueueBackByPath(path QueuePath, v any) error {
+	return t.rootNode.enqueueBackByPath(t, path, v)
+}
+
+// EnqueueFrontByPath enqueues an item in the front of the local queue of the Node
+// located at a given path through the TreeQueue; nodes for the path are created as needed.
+//
+// Enqueueing to the front is intended only for items which were first enqueued to the back
+// and then dequeued after reaching the front.
+//
+// Re-enqueueing to the front is only intended for use in cases where a queue consumer
+// fails to complete operations on the dequeued item, but failure is not yet final, and the
+// operations should be retried by a subsequent queue consumer. A concrete example is when
+// a queue consumer fails or disconnects for unrelated reasons while we are in the process
+// of dequeuing a request for it.
+//
+// path must be relative to the root node; providing a QueuePath beginning with "root"
+// will create a child node of root which is also named "root."
+func (t *TreeQueue) EnqueueFrontByPath(path QueuePath, v any) error {
+	return t.rootNode.enqueueFrontByPath(t, path, v)
+}
+
+func (t *TreeQueue) GetNode(path QueuePath) *Node {
+	return t.rootNode.getNode(path)
+}
+
+// Node maintains node-specific information used to enqueue and dequeue to itself, such as a local
+// queue, node depth, references to its children, and position in queue.
+// Note that the tenantQuerierAssignments DequeueAlgorithm largely disregards Node's queueOrder and
+// queuePosition, managing analogous state instead, because shuffle-sharding + fairness  requirements
+// necessitate input from the querier.
+type Node struct {
+	name             string
+	localQueue       *list.List
+	queuePosition    int      // next index in queueOrder to dequeue from
+	queueOrder       []string // order for dequeuing from self/children
+	queueMap         map[string]*Node
+	depth            int
+	dequeueAlgorithm DequeueAlgorithm
+	childrenChecked  int
+}
+
+func newNode(name string, depth int, da DequeueAlgorithm) (*Node, error) {
+	if da == nil {
+		return nil, fmt.Errorf("cannot create a node without a defined DequeueAlgorithm")
 	}
+	if tqa, ok := da.(*tenantQuerierAssignments); ok {
+		tqa.tenantOrderIndex = localQueueIndex - 1 // start from -2 so that we first check local queue
+		if tqa.tenantNodes == nil {
+			tqa.tenantNodes = map[string][]*Node{}
+		}
+	}
```

Comment:
This is a bit strange, any way to avoid leaking the details of tenantQuerierAssignments in here? As is, Nodes and TQA have a cyclic relationship setting each other's state, which is hard to follow IMO

Reply:
I agree, this isn't great -- one way to avoid this could be to define a `newTenantQuerierAssignments()` and only create tenant querier assignments with that, setting the expected defaults (-2 index, non-nil empty map), such that it's acceptable to expect non-nil values at this point. Does that sound reasonable?

Reply:
I don't think there's any reason we have to check the local queue first.

But to zoom out a bit, I think we can try to handle this similarly to how we do with the dequeue reporting if a node is empty after dequeue, where the tree basically reports "I created a new node" on enqueue and the selection algorithm updates in own state in response.

Reply:
^ was this meant to be a comment on this change? There's no local queue logic here.